File: back-end/src/main/java/pl/straburzynski/campfiresongs/song/exception/SongExistsException.java
package pl.straburzynski.campfiresongs.song.exception;

import com.google.common.collect.ImmutableMap;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.ResponseStatus;
import pl.straburzynski.campfiresongs.exception.CustomException;
import pl.straburzynski.campfiresongs.exception.ErrorResponse;
import pl.straburzynski.campfiresongs.song.model.Song;

@ResponseStatus(HttpStatus.BAD_REQUEST)
public class SongExistsException extends CustomException {

    public SongExistsException(Song song) {
        super(ErrorResponse.builder()
                .message("Song " + song.getTitle() + " by author " + song.getAuthor() + " already exists in database")
                .translationKey("exception.song_exists")
                .params(ImmutableMap.of("song", song))
                .build());
    }
}

File: back-end-kotlin/src/main/kotlin/pl/straburzynski/campfiresongs/song/exception/SongExistsException.kt
package pl.straburzynski.campfiresongs.song.exception

import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.ResponseStatus
import pl.straburzynski.campfiresongs.exception.CustomException
import pl.straburzynski.campfiresongs.exception.ErrorResponse
import pl.straburzynski.campfiresongs.song.model.SongDto


@ResponseStatus(HttpStatus.UNPROCESSABLE_ENTITY)
class SongExistsException(songDto: SongDto) : CustomException(
    ErrorResponse(
        mapOf("song" to songDto),
        "exception.song_exists",
        "Song " + songDto.title + " by author " + songDto.author + " already exists in database"
    ),
)